### PR TITLE
Validate recipe capabilities against a set of policies(ts):

### PR DESCRIPTION
**Pull request by mmandlis**
IngressValidation class introduces a method that examines all `create` handles, finds their restricted type (according to policies) and verifies each field's capabilities against the set of policies.

next steps:
(1) compute allowed capabilities for recipe stores that are not directly mentioned in policies (derived data)
(2) actually pass and enforce policies in recipe2plan 

b/160820832

### CODE REVIEW

File: src/runtime/policy/policy.ts [hunk 1]
```diff
@@ -15,7 +15,7 @@ import {Ttl, Capabilities, Capability, Persistence, CapabilityRange, Encryption}
 import {EntityType, InterfaceType, Type} from '../type.js';
 import {FieldPathType, resolveFieldPathType} from '../field-path.js';
 import {Handle} from '../recipe/handle.js';
-import {IngressValidationResult} from './ingress-validation.js';
+// import {IngressValidationResult} from './ingress-validation.js';
```
Comment (csilvestrini) [inline src/runtime/policy/policy.ts:18]: Delete?

File: src/runtime/policy/ingress-validation.ts [hunk 2]
```diff
@@ -14,37 +14,179 @@ import {Capability, Capabilities} from '../capabilities.js';
...
+      }
+    } else {
+      result.addError(
+        `Handle '${handle.id}' has no matching target type ` +
+            `${handle.type.resolvedType().toString()} in policies`);
+    }
+    return result;
+  }
+
+  private collectFieldPaths(fieldPrefix: string, fieldPaths: string[], fieldName: string, field) {
```
Comment (csilvestrini) [inline src/runtime/policy/ingress-validation.ts:99]: It looks like this uses fieldPaths as an output argument. Would this be clearer as a return value instead?

Usage is then:

```
let fieldPaths = [];
for (blah) {
  fieldPaths.push(...this.collectFieldPaths(blah));
}
```

File: src/runtime/policy/ingress-validation.ts [hunk 3]
```diff
@@ -14,37 +14,179 @@ import {Capability, Capabilities} from '../capabilities.js';
...
+
+      for (const fieldPath of fieldPaths) {
+        const fieldCapabilities = this.getFieldCapabilities(fieldPath);
+        assert(fieldCapabilities, `Missing capabilities for ${fieldPath}`);
+        const fieldResults = fieldCapabilities.map(
+            fc => fc.isAllowedForIngress(handle.capabilities));
+        if (!fieldResults.some(r => r.success)) {
+          result.addResult(IngressValidationResult.failWith(handle,
+              `Failed validating ingress for field '${fieldPath}'` +
+              ` of '${handle.id ? handle.id : handle.connections[0].getQualifiedName()}`,
```
Comment (csilvestrini) [inline src/runtime/policy/ingress-validation.ts:87]: Could be: `handle.id || handle.connections......`

File: src/runtime/policy/ingress-validation.ts [hunk 4]
```diff
@@ -14,37 +14,179 @@ import {Capability, Capabilities} from '../capabilities.js';
...
+          this.collectFieldPaths([fieldPrefix, fieldName].join('.'), fieldPaths, subfieldName, subfield);
+        }
+        break;
+      }
+      default:
+        assert(`Unsupported field kind: ${field.kind}`);
+    }
+  }
+
+  restrictType(type: Type): Type|null {
```
Comment (csilvestrini) [inline src/runtime/policy/ingress-validation.ts:121]: Comment for this method?

File: src/runtime/policy/ingress-validation.ts [hunk 5]
```diff
@@ -14,37 +14,179 @@ import {Capability, Capabilities} from '../capabilities.js';
...
+    for (const [fieldName, field] of Object.entries(type.getEntitySchema().fields)) {
+      const policyField = restrictedType.getEntitySchema().fields[fieldName];
+      if (policyField) {
+        fields[fieldName] = this.restrictField(field, policyField);
+      }
+    }
+    return EntityType.make([type.getEntitySchema().name], fields, type.getEntitySchema());
+  }
+
+  private restrictField(field, policyField) {
```
Comment (csilvestrini) [inline src/runtime/policy/ingress-validation.ts:134]: And this one, possibly?

File: src/runtime/policy/tests/ingress-validation-test.ts [hunk 6]
```diff
@@ -0,0 +1,413 @@
...
+    assert.isFalse(result.success);
+    assert.isTrue(result.toString().includes(expectedError),
+        `Expected ${result.toString()} to include ${expectedError}`);
+  };
+  const assertHandleIngressAllowed = async (manifestString, policy) => {
+    const recipe = (await Manifest.parse(manifestString)).recipes[0];
+    assert.isTrue(recipe.normalize());
+    assert.isTrue(recipe.isResolved());
+    const options = {errors: new Map()};
+    // const result = policy.isHandleIngressAllowed(recipe.handles[0]);
```
Comment (csilvestrini) [inline src/runtime/policy/tests/ingress-validation-test.ts:144]: ??

File: src/runtime/policy/policy.ts (hunk 1 above)
Comment (mmandlis) [reply src/runtime/policy/policy.ts]: done

File: src/runtime/policy/ingress-validation.ts (hunk 2 above)
Comment (mmandlis) [reply src/runtime/policy/ingress-validation.ts]: done. thanks, this is much nicer.

File: src/runtime/policy/ingress-validation.ts (hunk 3 above)
Comment (mmandlis) [reply src/runtime/policy/ingress-validation.ts]: done

File: src/runtime/policy/ingress-validation.ts (hunk 4 above)
Comment (mmandlis) [reply src/runtime/policy/ingress-validation.ts:126]: done

File: src/runtime/policy/ingress-validation.ts (hunk 5 above)
Comment (mmandlis) [reply src/runtime/policy/ingress-validation.ts:140]: done

File: src/runtime/policy/tests/ingress-validation-test.ts (hunk 6 above)
Comment (mmandlis) [reply src/runtime/policy/tests/ingress-validation-test.ts]: removed, this was a stale line

**Pull request merged**